Switched internal rotation to quaternion

DuccioLenkowicz opened the pull request:
Removed Rotation3 and EulerRotation3 properties to avoid singularity
issues in quaternion to euler conversion.
Added various methods to set internal rotation trough euler angles.
Euler angles are intendet as follows:
Yaw is applied first, roll second, pitch third.

DuccioLenkowicz committed:
Switched internal rotation to quaternion

Removed Rotation3 and EulerRotation3 properties to avoid singularity
issues in quaternion to euler conversion.
Added various methods to set internal rotation trough euler angles.
Euler angles are intendet as follows:
Yaw is applied first, roll second, pitch third.

rdeioris commented:
Can you expose SetEulerRotation and SetRotation as properties ? Transformations are semantically more near to an 'attribute' of an object instead of an 'action'

DuccioLenkowicz commented:
Sure i can do that.
I could also expose getters like the current Rotation3 and EulerRotation3 properties, and it would work, except if i write something like:
mesh.Rotation3 += new Vector3(0,0, someFloat);
In this case when the magnitude of the roll angle becomes greater than 90 degs, and i try to convert the quaternion back to euler, singularity comes in and the mesh gets stuck.

DuccioLenkowicz committed:
Exposed SetRotation via poperty

Added properties for SetRotation and SetEulerRotation in Mesh3 and
PerspectiveCamera.
Preliminarily added QuaternionToPitchYawRoll method to utils.

rdeioris commented:
This (rotation singularities) is something game developers need to grasp, so i suggest to expose both. In addition to them, a Rotate(x, y, z) method that rotates the mesh of the specified amount could be useful. Albeit the += form is what we are currently using, the .Rotate() one is the most common in other engines.

DuccioLenkowicz committed:
Revert "Exposed SetRotation via poperty"

This reverts commit df8fb09d5ff79d852ec203ef9958746775431733.
